lib/LaTeXML/Package/listings.sty.ltxml

```diff
@@ -118,16 +118,26 @@ DefMacroI(T_CS('\begin{lstlisting}'), 'OptionalKeyVals:LST', sub {
     lstActivate($keyvals);
     return lstProcessDisplay(lstGetTokens('name'), $text); });
 
-DefMacro('\lstinputlisting OptionalKeyVals:LST Semiverbatim', sub {
+DefParameterType('LstNameVerbatim', sub {
+    my ($gullet) = @_;
+    # Closer to listings approch, but do NOT make - => \textendash! (ends up in file name)
+    AssignCatcode('$', CC_ACTIVE);
+    AssignCatcode('_', CC_ACTIVE);
+    DefMacroI(T_ACTIVE('$'), undef, T_CS('\textdollar'),     scope => 'local');
+    DefMacroI(T_ACTIVE('_'), undef, T_CS('\textunderscore'), scope => 'local');
```

Comment:
Does this work with
```latex
\lxWithClass{class}{\lstinputlisting{a_b}}
```
? I gather `\lxWithClass` will prevent `LstNameVerbatim` from affecting the catcode and so you get the wrong name again. In this respect, `\lstinputlisting` is a robust command.

Reply:
Yeah, right, but then *all* the proposed solutions suffer this, don't they?  `\lstinputlisting` being robust (which it apparently isn't) doesn't really affect this either, since it's the argument which already will have been read, rather than delaying the expansion of `\lstinputlisting`.

Digging in the `listings` code, I see that we've been barking up the wrong tree; Listings isn't changing catcodes (at least not for `$_-`) before reading the filename, but rather it's substituting for them before using it in the caption!  Perversely, it's much like using `\texttt` (or a non-typewriter `\texttt`)!

Reply:
> _all_ the proposed solutions suffer this, don't they?

Actually my PR does exactly what `\lstinputlisting` does, and should survive `\lxWithClass` (I admit I haven't checked). Well, I left out the `&ndash;` because it would end up in `@download`, and a minor visual discrepancy felt ok compared to saving the original name in a different variable (which would be `\lst@intname` if we follow the original code).

(My PR also does *some* catcode handling related to `=` due to babel activating `=`, which I find totally mysterious. Also a `\makeatletter` before reading the keyvals, because... uhm, so that the key names can contain `@` as letter? The listings package does some weird stuff.)

Reply:
Indeed, the substitution was the effect; but burying it in `lstActivate` confused me.

Reply:
> burying it in `lstActivate` confused me.

I suspect I wanted to avoid the thing I mention below: listings reads the file whose name is the string of characters received from the mouth, *then* edits and saves the filename. So doing this directly in KeyVals is too early, especially if you decide you actually want the en dashes after all.

Reply:
> whose name is the string of characters received from the mouth

What did I just say... it's the _expansion_ of the string received from the mouth. The point being that listings has two distinct logical names: the physical filename, and the TOC filename, and using a KeyVals type forces them to coincide.